turbopack/crates/turbo-tasks/src/scope.rs

```diff
@@ -1,52 +1,230 @@
-use std::sync::Arc;
+//! Parallel for each resp. map running in the current tokio thread pool maintaining turbo tasks and
+//! tracing context.
+//!
+//! This avoid the problem of sleeping threads with mimalloc when using rayon in combination with
+//! tokio. It also avoid having multiple thread pools.
 
-use crate::{TurboTasksApi, turbo_tasks, turbo_tasks_scope};
+use std::{
+    marker::PhantomData,
+    mem::take,
+    panic::{self, AssertUnwindSafe, catch_unwind},
+    pin::Pin,
+    sync::{
+        Arc,
+        atomic::{AtomicUsize, Ordering},
+    },
+};
 
-/// A wrapper around [`rayon::Scope`] that preserves the [`turbo_tasks_scope`].
-pub struct Scope<'scope, 'a> {
-    scope: &'a rayon::Scope<'scope>,
-    handle: tokio::runtime::Handle,
-    turbo_tasks: Arc<dyn TurboTasksApi>,
-    span: tracing::Span,
+use parking_lot::Mutex;
+use tokio::{
+    runtime::{Builder, Handle},
+    task::{JoinHandle, block_in_place},
+};
+use tracing::{Instrument, Span};
+
+use crate::{
+    TurboTasksApi,
+    manager::{try_turbo_tasks, turbo_tasks_future_scope},
+};
+
+/// Scope to allow spawning tasks with a limited lifetime.
+///
+/// Dropping this Scope will wait for all tasks to complete.
+pub struct Scope<'scope, 'env: 'scope, R: Send + 'env> {
+    results: &'scope [Mutex<Option<R>>],
+    index: AtomicUsize,
+    futures: Mutex<Vec<JoinHandle<()>>>,
+    handle: Handle,
+    turbo_tasks: Option<Arc<dyn TurboTasksApi>>,
+    span: Span,
+    /// Invariance over 'env, to make sure 'env cannot shrink,
+    /// which is necessary for soundness.
+    ///
+    /// see https://doc.rust-lang.org/src/std/thread/scoped.rs.html#12-29
+    env: PhantomData<&'env mut &'env ()>,
 }
 
-impl<'scope> Scope<'scope, '_> {
-    pub fn spawn<Body>(&self, body: Body)
+impl<'scope, 'env: 'scope, R: Send + 'env> Scope<'scope, 'env, R> {
+    fn new(results: &'scope [Mutex<Option<R>>]) -> Self {
```

Comment:
Nit: Let's make this `unsafe`, as it requires that the caller drop and not forget the `Scope` object.
```suggestion
    /// Safety: The caller must ensure `Scope` is dropped and not forgotten.
    unsafe fn new(results: &'scope [Mutex<Option<R>>]) -> Self {
```